projects/core/src/auth/user-auth/guards/custom-login.guard.ts

```diff
@@ -0,0 +1,111 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { inject, Injectable } from '@angular/core';
+import { CanActivate, GuardResult, Router } from '@angular/router';
+import { catchError, map, Observable, of, tap } from 'rxjs';
+import {
+  GlobalMessageService,
+  GlobalMessageType,
+} from '../../../global-message';
+import { SemanticPathService } from '../../../routing/configurable-routes/url-translation/semantic-path.service';
+import { StorageSyncType } from '../../../state/config/state-config';
+import {
+  getStorage,
+  persistToStorage,
+  readFromStorage,
+} from '../../../state/utils/browser-storage';
+import { WindowRef } from '../../../window/window-ref';
+import { AuthService } from '../facade/auth.service';
+import { CsrfStateService } from '../facade/csrf-state.service';
+
+const STORAGE_KEY = 'login_redirect_count';
+const timeout = 15_000;
+const totalRetries = 2;
+
+interface CustomLoginGuardMetadata {
+  /** Add timeout to recover from stale/interrupted state */
+  t: number;
+  /** Redirect count */
+  c: number;
+}
+
+/**
+ * This guard requests the CSRF token required for the custom login form as a way
+ * to check for a valid Authorization Server session.
+ *
+ * It will redirect the user to restart the auth flow on failed requests, keeping
+ * track of the redirects to avoid creating an infinite loop of redirects.
+ */
+@Injectable({
+  providedIn: 'root',
+})
+export class CustomLoginGuard implements CanActivate {
+  authService = inject(AuthService);
+  router = inject(Router);
+  semanticPathService = inject(SemanticPathService);
+  windowRef = inject(WindowRef);
+  storage = getStorage(StorageSyncType.LOCAL_STORAGE, this.windowRef);
+  globalMessageService = inject(GlobalMessageService);
+  csrfStateService = inject(CsrfStateService);
+
+  canActivate(): Observable<GuardResult> {
+    return this.authService.getCsrfToken().pipe(
+      tap((token) => {
+        this.csrfStateService.set(token);
+        this.clearRedirectCount();
+      }),
+      map(() => true),
+      catchError(() => {
+        const currentCount = this.getRedirectCount();
+        if (currentCount >= totalRetries) {
+          // retry limit met, go to homepage with message
+          this.clearRedirectCount();
+          this.globalMessageService.add(
+            { key: 'authMessages.unrecoverableError' },
+            GlobalMessageType.MSG_TYPE_ERROR
+          );
+          return of(this.createRoute('home'));
+        }
+        this.setRedirectCount(currentCount + 1);
+
+        // retry until limit met
+        return of(this.createRoute('login'));
+      })
+    );
+  }
+
+  getRedirectCount() {
+    const countMeta = readFromStorage<CustomLoginGuardMetadata>(
+      this.storage as Storage,
+      STORAGE_KEY
+    );
+    if (countMeta && Date.now() - countMeta.t < timeout) {
+      return countMeta.c;
+    }
+    return 0;
+  }
+
+  setRedirectCount(count: number) {
+    return persistToStorage(
```

Comment:
if we don't use `StatePersistenceService`, our field looks like `login_redirect_count` instead of 
`spartacus⚿login_redirect_count`. Not sure if it's standard

But anyway, I don't see other usages of those functions outside 1StatePersistenceService1 in the whole project. What benefits do we have over `this.winRef.localStorage?.setItem(STORAGE_KEY, { t: Date.now(), c: count })` ?

Reply:
I used the storage methods for simplicity, though it might have been better to use localStorage directly instead of the utility method.  I don't know if there is a standard.

StatePersistenceService also requires the value to be passed as an observable, and while it should be synchronous since there are no async schedulers involved, it could become an issue later if the write operation is scheduled for a task then canceled because of browser navigation.